### DISCUSSION TITLE
Cypress log out after 3DS check for checkout.com

### DISCUSSION THREAD

**Discussion by umerzaman** [cypress-io/cypress#26164] [Questions and Help]
Im writing a test for the payment section on Cypress after clicking on the payment button page goes to the external URL. e.g. from http://localhost:64574/ to "https://3ds2-sandbox.ckotech.co/interceptor/3ds_ibgzskqjjfqevj4w3gwmdptp4m/device-information"

after completing the 3ds check the page to redirect back and went to the login page. I'm wrapping my code in the cy.session you can see below.

`beforeEach(() => { 

        cy.fixture('example.json').then(function(data)
        {
            configdata=data  
        })
        cy.fixture('3dsCard.json').then(function(data)
        {
            configdata3DS=data  
        })
        cy.session('session', () => {
            cy.visit(Cypress.env('localURL'))
            cy.get('#onetrust-accept-btn-handler').click();
            cy.get('#VoucherNo').type('myvoucher-no')
            cy.get('#PinCode').type('my-pin')
            cy.get('.btn').click() 
            cy.getCookie('ASP.NET_SessionId')
            .should('exist'),
            { 
                cacheAcrossSpecs: true
              }
          }) 
    })`

I don't know what's going on in the cypress test environment locally I'm testing this with a card it's working fine I can see the confirmation page but in the test goes back to login.

**Comment by portyu9** [comment]
Hi @umerzaman 

There are two separate things worth fixing/checking here.

First, your `cy.session()` options are in the wrong place. In the code shown, this:

```js
{
  cacheAcrossSpecs: true
}
```

is inside the setup callback. `cy.session()` expects options as its **third argument**:

```js
cy.session(
  'session',
  () => {
    cy.visit(/* login page */)

    // perform login...
  },
  {
    validate() {
      cy.getCookie('ASP.NET_SessionId').should('exist')
    },
    cacheAcrossSpecs: true,
  }
)
```

That said, `cacheAcrossSpecs` only controls reuse between spec files. It does **not** by itself explain why the application returns from 3DS to the login page.

I would diagnose the authentication state around the 3DS flow:

```js
cy.getCookie('ASP.NET_SessionId').should('exist')

// start 3DS...

// after returning to your application
cy.location('origin').should('eq', 'http://localhost:64574')
cy.getCookie('ASP.NET_SessionId').should('exist')
```

If the cookie disappears, investigate its `Domain`, `SameSite`, `Secure`, and callback/navigation behavior.

If the cookie is still present but the server redirects to login, then the server/application is rejecting the session after the 3DS callback; inspect that callback request and response.

Also, with current Cypress, if the 3DS flow performs **top-level navigation** to another origin and you need to execute Cypress commands there, those commands belong inside `cy.origin()`. If the provider challenge is instead inside a cross-origin iframe, that is a different limitation.

So I would fix the `cy.session()` syntax, then determine whether the session cookie is actually lost before assuming Cypress is logging the user out.

Please mark this answer as accepted if it helped, thank you.
